bootstrap.sh

```diff
@@ -0,0 +1,205 @@
+#!/bin/sh
+
+#
+# Bootstrap 0.1.0 for reMarkable (rM2 untested)
+# wget -qO- '<URL_to_bootstrap.sh>' | sh -s [packages]...
+#
+
+bootstrap_wget() {
+  if [ ! -d /home/root/.cache/wget_bin ]
+  then
+      log INFO "Installing a current wget binary with better and current tls support" \
+          "to /home/root/.cache/wget_bin/ for this installation..." \
+          ""
+      # Bootstrap a current version of wget
+      WGET_BINARIES_PATH='http://static.cosmos-ink.net/remarkable/artifacts'
```

Comment:
where is this hosted?

Reply:
On my Raspberry Pi 4 at home:
- Nginx on Ubuntu Server 20.04
- 30 MBit/s Uplink
- Automatic updating of DNS entries should the ISP have problems (was lately down about 3 times a week for about 3 mins with new ip afterwards)

I can also easily move that to a VPS I have. Or better yet, we just host it under thirdparty on the toltec server.

The file was [built here](https://git.cosmos-ink.net/linus/wget-remarkable-pipeline). You can check out the latest artifact from the CI/CD section there. Should be same hash and commit and build logs are also all there.

Reply:
I'd feel much better if it was hosted somewhere that had redundancy.

Reply:
Is there a reason for not using the [wget package from Entware](http://bin.entware.net/armv7sf-k3.2/wget_1.20.3-4_armv7-3.2.ipk)?

Reply:
To install entware (mostly atm) from https in the first place.

Reply:
Here’s an alternative idea: we could run the install script from the context of a host connected to the rM through SSH. That way, we could leverage the up-to-date wget/curl from the host. I believe this would make things simpler. What do you think?

Reply:
In theory yes, but then we would also need to ensure that user has wget installed. Either downloading it on Windows (and installing on Mac) or requiring docker for just one small installation.  
Considering most end users don't run linux, this would be a hassle for both us and them.

Reply:
Since, the initial `wget -qO- https://static.cosmos-ink.net/remarkable/bootstrap.sh | sh -s` invocation uses the insecure default `wget` from the reMarkable, doesn’t that make all further efforts towards avoiding MITM moot? It seems to me that an attacker can simply MITM the bootstrap.sh file in this case.

Reply:
We should include a hash check in that command. Instead of just piping it into sh.

Reply:
Agreed. That's the dilemma I see currently: Short one-liner vs secure script snippet. We should really go with the latter though.

Reply:
I hope that the reMarkable team notices this and gives their BusyBox a update (wget is part of that). Then we would use a longer script for about 2 months and could switch to a oneliner afterwards. We could also get rid of any wget bootstrapping.

Reply:
The secure snippet wouldn't really be that long though.

Reply:
What about embedding the secure wget binary inside the `bootstrap` script? This way we 1) remove an additional HTTP request, and 2) only need to check the SHA-sum of the downloaded script, instead of both the script and the secure wget binary retrieved by the script.

Reply:
The embedding could be done automatically by a build script, and the final `bootstrap` script could be automatically published to the Toltec server.

Reply:
> What about embedding the secure wget binary inside the `bootstrap` script? This way we 1) remove an additional HTTP request, and 2) only need to check the SHA-sum of the downloaded script, instead of both the script and the secure wget binary retrieved by the script.
> 
> The embedding could be done automatically by a build script, and the final `bootstrap` script could be automatically published to the Toltec server.

We could probably do that. This would make the script fairly big though. And the fact that the script already has the hash of the weg file builds a chain of trust in that case anyway. If bootstrap.sh has a known hash, that would change if we were to change the hash in that script. It's similar in essence to how git commits work as well, if I'm not mistaken.